### PR TITLE
Added support for dry-runs to weather-mv tool

### LINKED ISSUE #22: Add support for dry-runs to `weather-mv`.
A helpful step towards fixing #21.

_ticket updated: 2022-03-24_

As a weather tools user, I would like to be able to preview the effect of each tool before incurring the cost of data movement and infrastructure. These light-weight previews will help me test pipelines before deployment, lowering the number of iterations needed to set up a data pipeline. For this issue, I want to be able to perform dry runs with the weather mover.

## Acceptance Criteria
* Provide a common interface for exercising dry runs for every Data Sink 
* When a user passes the `-d` or `--dry-run` flag to the `weather-mv` cli, this feature will be activated. 
* When a user checks tool documentation (the README or CLI help message), they will have a good understanding of what the feature does
* As a user, I will still have some way to monitor the execution flow of the tool during a dry run
  * Log messages from non-dry runs will remain the same as those within a dry-run
  * As a user, I can inspect the kinds of messages that _would_ have been written to BigQuery. 
  * (optional) Maybe more log messages are needed to see what's happening?
* As a user, I can execute dry runs locally or remotely on Dataflow
* Where appropriate, data is simulated in memory. No data is written to disk or cloud storage during a dry run. 
  * As a user, I still would like to validate execution on actual user-suppled URIs.
* Where there are contradictions in requirements, the ergonomic option for `weather-mv` users is preferred.    
* All code should be completely covered by tests. 

## Implementation Notes
* The best place to provide a common interface for dry runs is https://github.com/google/weather-tools/blob/fd0c5e4f63b2bebea30d743e1f04f9c008ce914f/weather_mv/loader_pipeline/sinks.py#L33
* Most code changes will happen in this class https://github.com/google/weather-tools/blob/fd0c5e4f63b2bebea30d743e1f04f9c008ce914f/weather_mv/loader_pipeline/bq.py#L41

### PR DESCRIPTION
**Pull request by mahrsee1997**
Fixed #22.

### CODE REVIEW

File: weather_mv/loader_pipeline/bq.py [hunk 1]
```diff
@@ -82,32 +82,44 @@ def __post_init__(self):
...
+                        project=self.table.project,
+                        dataset=self.table.dataset_id,
+                        table=self.table.table_id,
+                        write_disposition=BigQueryDisposition.WRITE_APPEND,
+                        create_disposition=BigQueryDisposition.CREATE_NEVER)
+            )
+        else:
+            (
+                extracted_rows
+                | "Log Extracted Rows" >> beam.Map(logger.debug)
```
Comment (alxmrs) [inline weather_mv/loader_pipeline/bq.py:121]: In a separate issue, we need to update the CLIs for all the weather tools s.t. users can set the log level to debug. Would you mind filing a GH issue for this?

File: weather_mv/README.md [hunk 2]
```diff
@@ -51,6 +50,16 @@ weather-mv --uris "gs://your-bucket/*.nc" \
...
+Preview load with a dry run:
+
+```bash
+weather-mv --uris "gs://your-bucket/*.nc" \
+           --output_table $PROJECT.$DATASET_ID.$TABLE_ID \
+           --temp_location "gs://$BUCKET/tmp" \  # Needed for batch writes to BigQuery
+           --direct_num_workers 2 \
+           --dry-run
+```
+
```
[the remark this answers is not in the document]
Comment (alxmrs) [reply weather_mv/README.md:62]: Sanity check: Have you tested this e2e locally? Does it work with just these changes?

**Review by alxmrs — COMMENTED** [review]
Left a few comments; overall this looks good.

File: weather_mv/loader_pipeline/bq.py (hunk 1 above)
Comment (mahrsee1997) [reply weather_mv/loader_pipeline/bq.py:121]: Created GH issue #135 for the same.

File: weather_mv/loader_pipeline/bq.py [hunk 3]
```diff
@@ -82,32 +82,44 @@ def __post_init__(self):
...
+        if not self.dry_run:
+            # Create the table in BigQuery
+            try:
+                table = bigquery.Table(self.output_table, schema=table_schema)
+                self.table = bigquery.Client().create_table(table, exists_ok=True)
+            except Exception as e:
+                logger.error(f'Unable to create table in BigQuery: {e}')
+                raise
+        else:
+            logger.debug("Created the BigQuery table")
```
[the remark this answers is not in the document]
Comment (mahrsee1997) [reply weather_mv/loader_pipeline/bq.py:94]: Yes, it makes sense. I tried out few ways to display schema to the user. What do you think about these ?
Please find the attached screenshots for your reference.

1. Using pprint - Data pretty printer
Approach 1 - pformat

2. Using bigquery.Client().schema_to_json()
Approach 2

3. Custom logic by iterating over `table_schema `
Approach 3

File: weather_mv/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (mahrsee1997) [reply weather_mv/README.md:62]: Yes i checked it and it works fine.

Screenshot from 2022-03-31 19-02-32

File: weather_mv/loader_pipeline/bq.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alxmrs) [reply weather_mv/loader_pipeline/bq.py:94]: These experiments are great! I like option 1 the best. It should definitely be in one long message to make it easier to read from cloud log tracing products. Option 2 is good, but optimized for machines and not human readability.

[the remark this answers is not in the document]
Comment (mahrsee1997) [reply weather_mv/loader_pipeline/bq.py]: Yes, I too think option 1 is the best. Pushed the changes for the same.

File: weather_mv/loader_pipeline/bq.py [hunk 4]
```diff
@@ -82,32 +82,44 @@ def __post_init__(self):
...
                 table_schema = dataset_to_table_schema(ds)
 
-        # Create the table in BigQuery
-        try:
-            table = bigquery.Table(self.output_table, schema=table_schema)
-            self.table = bigquery.Client().create_table(table, exists_ok=True)
-        except Exception as e:
-            logger.error(f'Unable to create table in BigQuery: {e}')
-            raise
+        if not self.dry_run:
```
[the remark this answers is not in the document]
Comment (mahrsee1997) [reply weather_mv/loader_pipeline/bq.py]: Necessary code changes completed and pushed.

**Review by alxmrs — APPROVED** [review]
LGTM

**Pull request merged**